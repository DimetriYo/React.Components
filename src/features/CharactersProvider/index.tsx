import { PropsWithChildren, createContext, useContext, useEffect, useState } from 'react';
import { TCharacter } from '../../types';
import { useSearchParams } from 'react-router-dom';
import { fetchCharacters } from '../../api/fetchCharacters';

export const CharactersContext = createContext<TCharacter[]>([]);
export const PagesCountContext = createContext<number>(0);

export const useCharacters = () => useContext(CharactersContext);
export const usePagesCount = () => useContext(PagesCountContext);

export function CharactersProvider({ children }: PropsWithChildren) {
  const [characters, setCharacters] = useState<TCharacter[]>([]);
  const [pagesCount, setPagesCount] = useState<number>(0);
  const [searchParams] = useSearchParams();

  const page = searchParams.get('page') || undefined;
  const searchTerm = searchParams.get('searchTerm') || undefined;

  useEffect(() => {
    fetchCharacters({ page, searchTerm }).then(({ info: { pages }, results }) => {
      setCharacters(results);
      setPagesCount(pages);
    });
  }, [page, searchTerm]);

  return (
    <CharactersContext.Provider value={characters}>
      <PagesCountContext.Provider value={pagesCount}>{children}</PagesCountContext.Provider>
    </CharactersContext.Provider>
  );
}
