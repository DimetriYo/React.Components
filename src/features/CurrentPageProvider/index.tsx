import {
  PropsWithChildren,
  createContext,
  useContext,
  useEffect,
  useState,
} from 'react';
import { useSearchParams } from 'react-router-dom';

const CurrentPaqeContext = createContext<[number, (x: number) => void]>([
  0,
  () => {},
]);
export const useCurrentPage = () => useContext(CurrentPaqeContext);

export function CurrentPageProvider({ children }: PropsWithChildren) {
  const [currentPage, setCurrentPage] = useState<number>(0);
  const [searchParams, setSearchParams] = useSearchParams();
  const urlPage = searchParams.get('page');

  const updateCurrentPage = (newPageNumber: number) => {
    setCurrentPage(newPageNumber);
    setSearchParams((oldSP) => {
      oldSP.set('page', newPageNumber.toString());

      return oldSP;
    });
  };

  useEffect(() => {
    const urlPageNumber = Number(urlPage);
    if (urlPage !== null) {
      if (Number.isInteger(urlPageNumber)) {
        setCurrentPage(urlPageNumber);
      } else {
        setSearchParams((oldSP) => {
          oldSP.delete('page');

          return oldSP;
        });
      }
    }
  }, []);

  return (
    <CurrentPaqeContext.Provider value={[currentPage, updateCurrentPage]}>
      {children}
    </CurrentPaqeContext.Provider>
  );
}
