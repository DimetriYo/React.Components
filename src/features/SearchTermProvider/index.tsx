import { PropsWithChildren, createContext, useContext, useState } from 'react';

const SearchTermContext = createContext('');
const SetSearchTermContext = createContext<(newSearchTerm: string) => void>(() => {});

export function SearchTermProvider({ children }: PropsWithChildren) {
  const oldSearchTerm = localStorage.getItem('searchTerm');
  const [searchTerm, setSearchTerm] = useState(oldSearchTerm || '');
  const updateSearchTerm = (newSearchTerm: string) => {
    localStorage.setItem('searchTerm', newSearchTerm);
    setSearchTerm(newSearchTerm);
  };

  return (
    <SearchTermContext.Provider value={searchTerm}>
      <SetSearchTermContext.Provider value={updateSearchTerm}>
        {children}
      </SetSearchTermContext.Provider>
    </SearchTermContext.Provider>
  );
}

export const useSearchTerm = () => useContext(SearchTermContext);
export const useSetSearchTerm = () => useContext(SetSearchTermContext);
