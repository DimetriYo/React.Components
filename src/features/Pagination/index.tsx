import { useState } from 'react';
import { PaginationCircle } from './PaginationCircle';
import style from './style.module.css';
import { useCharacters } from '../CharactersProvider';
import { useSearchTerm } from '../SearchTermProvider';

export function Pagination() {
  const [currentPage, setCurrentPage] = useState<number>(1);
  const [, pagesCount, updateCharacters] = useCharacters();
  const [searchTerm] = useSearchTerm();

  const handlePageBtnClick = (page: number) => {
    setCurrentPage(page);
    updateCharacters(page, searchTerm || '');
  };

  return (
    pagesCount > 0 && (
      <div className={style.pagesList}>
        <PaginationCircle index={currentPage.toString()} active />
        {Array(pagesCount)
          .fill(null)
          .map((_, index) => (
            <PaginationCircle
              index={(index + 1).toString()}
              key={index}
              onClick={() => handlePageBtnClick(index + 1)}
            />
          ))}
      </div>
    )
  );
}
