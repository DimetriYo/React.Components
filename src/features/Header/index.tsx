import { DetailedHTMLProps, HTMLAttributes } from 'react';
import { Pagination } from '../Pagination';
import { SearchField } from './components/SearchField';

type Props = DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>;

export function Header(props: Props) {
  return (
    <header {...props}>
      <SearchField />
      <Pagination />
    </header>
  );
}
