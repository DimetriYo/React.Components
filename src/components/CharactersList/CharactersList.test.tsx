import { afterEach, beforeAll, describe, expect, it, vi } from 'vitest';
import { cleanup, render, screen } from '@testing-library/react';
import { useCharacters } from '../../features/CharactersProvider';
import { CharactersList } from '.';
import { MemoryRouter } from 'react-router-dom';

afterEach(() => {
  vi.clearAllMocks();
  cleanup();
});

beforeAll(() => {
  vi.mock('../../features/CharactersProvider', () => ({
    useCharacters: vi.fn(),
  }));
});

describe('Characters list', () => {
  it('shold render empty page if there are no cards', () => {
    vi.mocked(useCharacters).mockReturnValue([]);
    render(<CharactersList />);
    const component = screen.getByRole('heading');
    expect(component).toHaveTextContent('Sorry');
  });

  it('shold render 1 card', () => {
    vi.mocked(useCharacters).mockReturnValue([
      {
        id: 0,
        name: 'TestChar',
        status: 'Alive',
        species: '',
        type: '',
        gender: '',
        origin: {},
        location: {},
        image: '',
        episode: ['1'],
        url: '',
      },
    ]);

    const component = render(
      <MemoryRouter>
        <CharactersList />
      </MemoryRouter>
    );

    expect(component).toMatchSnapshot();
  });
});
