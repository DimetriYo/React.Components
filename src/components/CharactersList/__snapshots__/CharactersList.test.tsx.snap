// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Characters list > shold render 1 card 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="max-h-full overflow-y-scroll flex-grow basis-1/2"
      >
        <div
          class="flex gap-4 flex-wrap items-center justify-center p-4"
        >
          <a
            href="/0"
          >
            <figure
              class="_characterCard_ce7192"
              data-character-card="true"
            >
              <img
                alt="TestChar photo"
                src=""
              />
              <figcaption
                class="text-center"
                data-testid="character-name"
              >
                TestChar
              </figcaption>
            </figure>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="max-h-full overflow-y-scroll flex-grow basis-1/2"
    >
      <div
        class="flex gap-4 flex-wrap items-center justify-center p-4"
      >
        <a
          href="/0"
        >
          <figure
            class="_characterCard_ce7192"
            data-character-card="true"
          >
            <img
              alt="TestChar photo"
              src=""
            />
            <figcaption
              class="text-center"
              data-testid="character-name"
            >
              TestChar
            </figcaption>
          </figure>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
